src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs

```diff
@@ -557,6 +554,43 @@ private bool GetCanVersionIncremental(IncrementalStatus buildInfoIncrementalStat
             return true;
         }
 
+        private IEnumerable<FileItem> GetFilesToCalculateAttributes()
+        {
+            var keys = new HashSet<string>();
+            foreach (var f in _parameters.Files.EnumerateFiles())
+            {
+                var fileKey = ((RelativePath)f.File).GetPathFromWorkingFolder().ToString();
+                keys.Add(fileKey);
+                yield return new FileItem
+                {
+                    PathFromWorkingFolder = fileKey,
+                    FullPath = f.FullPath,
+                    IsFromSource = true,
+                };
+            }
+            if (CurrentBuildVersionInfo.Dependency != null)
+            {
+                foreach (var f in CurrentBuildVersionInfo.Dependency.GetAllDependentNodes())
+                {
+                    var p = RelativePath.TryParse(f);
+                    if (p == null)
```

Comment:
if we validate dependencies when saving them, will it make the dependency graph smaller?